src/Compilers/CSharp/Test/Emit2/Semantics/RefReadonlyParameterTests.cs

```diff
@@ -1251,4 +1251,691 @@ static class C
+    [Fact]
+    public void RefReadonlyParameter_RefArgument_OverloadResolution_01()
+    {
+        var source = """
+            class C
+            {
+                static string M1(string s, ref int i) => "string" + i;
+                static string M1(object o, in int i) => "object" + i;
+                static string M1(C c, ref readonly int i) => "c" + i;
+                static void Main()
+                {
+                    int i = 5;
+                    System.Console.WriteLine(M1(null, ref i));
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (9,34): error CS0121: The call is ambiguous between the following methods or properties: 'C.M1(string, ref int)' and 'C.M1(C, ref readonly int)'
+            //         System.Console.WriteLine(M1(null, ref i));
+            Diagnostic(ErrorCode.ERR_AmbigCall, "M1").WithArguments("C.M1(string, ref int)", "C.M1(C, ref readonly int)").WithLocation(9, 34));
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_RefArgument_OverloadResolution_01_Ctor()
+    {
+        var source = """
+            class C
+            {
+                private C(string s, ref int i) => System.Console.WriteLine("string" + i);
+                private C(object o, in int i) => System.Console.WriteLine("object" + i);
+                private C(C c, ref readonly int i) => System.Console.WriteLine("c" + i);
+                static void Main()
+                {
+                    int i = 5;
+                    new C(null, ref i);
+                }
+            }
+            """;
+        CreateCompilation(source).VerifyDiagnostics(
+            // (9,13): error CS0121: The call is ambiguous between the following methods or properties: 'C.C(string, ref int)' and 'C.C(C, ref readonly int)'
+            //         new C(null, ref i);
+            Diagnostic(ErrorCode.ERR_AmbigCall, "C").WithArguments("C.C(string, ref int)", "C.C(C, ref readonly int)").WithLocation(9, 13));
+    }
+
+    [Fact]
+    public void RefReadonlyParameter_RefArgument_OverloadResolution_02()
+    {
+        var source = """
+            class C
+            {
+                static string M1(string s, ref int i) => "string" + i;
+                static string M1(object o, in int i) => "object" + i;
+                static string M1(C c, ref readonly int i) => "c" + i;
+                static void Main()
+                {
+                    int i = 5;
+                    System.Console.WriteLine(M1(default(string), ref i));
+                    System.Console.WriteLine(M1(default(object), ref i));
+                    System.Console.WriteLine(M1(default(C), ref i));
+                }
+            }
+            """;
+        CompileAndVerify(source, expectedOutput: """
+            string5
+            object5
+            c5
+            """).VerifyDiagnostics(
+            // (10,58): warning CS9502: Argument 2 should not be passed with the 'ref' keyword
```

Comment:
>should not be passed with the 'ref' keyword [](http://example.com/codeflow?start=51&length=43)

Consider adjusting the wording so that it explains the effect of using 'ref' keyword and provides guidance how to fix the warning. For example, if compiler "interprets" 'ref' at this call site as 'in', we could say something like: "The 'ref' modifier for an argument corresponding to 'ref' parameter is equivalent to 'in'. Consider using it instead."